### PR TITLE
 [clang][InstallAPI] Introduce basic driver to write out tbd files

### PR DESCRIPTION
**Pull request by cyndyishida**
This introduces a basic outline of installapi as a clang driver option. It captures relevant information as cc1 args, which are common arguments already passed to the linker to encode into TBD file outputs. This is effectively an upstream for what already exists as `tapi installapi` in Xcode toolchains, but directly in Clang. This patch does not handle any AST traversing on input yet.

InstallAPI is broadly an operation that takes a series of header files that represent a single dynamic library and generates a TBD file out of it which represents all the linkable symbols and necessary attributes for statically linking in clients. It is the linkable object in all Apple SDKs and when building dylibs in Xcode. `clang -installapi` also will support verification where it compares all the information recorded for the TBD files against the already built binary, to catch possible mismatches like when a declaration is missing a definition for an exported symbol.

### CODE REVIEW

**Comment by cyndyishida** [comment]
In case anyone is interested, 
Here are relevant links for upstreaming `tapi` in LLVM: 
* [Inital RFC](https://discourse.llvm.org/t/rfc-open-sourcing-and-contributing-tapi-back-to-the-llvm-community/46214/24)
* [Continued RFC](https://discourse.llvm.org/t/rfc-continuation-of-tapi-into-llvm/52825)
* [Dev Talk](https://www.youtube.com/watch?v=B9li6EkD5zA)
* [Full TAPI sources](https://github.com/apple-oss-distributions/tapi), includes docs

**Review by ributzka — COMMENTED** [review]
It is great to finally see this come together :D Thank you Cyndy

File: clang/include/clang/InstallAPI/Context.h [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  std::unique_ptr<llvm::raw_pwrite_stream> OS = nullptr;
+
+  /// DiagnosticsEngine to report errors.
+  llvm::IntrusiveRefCntPtr<DiagnosticsEngine> Diags = nullptr;
+
+  /// File Path of output location.
+  StringRef OutputLoc{};
+
+  /// What encoding to write output as.
+  llvm::MachO::FileType FT = llvm::MachO::FileType::TBD_V5;
```
[the remark this answers is not in the document]
Comment (cyndyishida) [reply clang/include/clang/InstallAPI/Context.h:48]: Yes, `v5` is the default but will be configurable, I added it here now to layout the correct flow for emitting the TBD.

File: clang/lib/Driver/Driver.cpp [hunk 2]
```diff
@@ -4319,6 +4324,12 @@ void Driver::BuildActions(Compilation &C, DerivedArgList &Args,
     if (!MergerInputs.empty())
       Actions.push_back(
           C.MakeAction<IfsMergeJobAction>(MergerInputs, types::TY_Image));
+  } else if (Args.hasArg(options::OPT_installapi)) {
+    assert(Inputs.size() == 1 && "InstallAPI action can only handle 1 input");
```
[the remark this answers is not in the document]
Comment (cyndyishida) [reply clang/lib/Driver/Driver.cpp:4328]: This is temporary and will be lifted when we can handle multiple inputs. Worth noting though, that I'm going to keep the original `tapi` input style where we only accept directory structures or JSON input.

[the remark this answers is not in the document]
Comment (ributzka) [reply clang/lib/Driver/Driver.cpp:4328]: Could you please add a comment or TODO?

**Review by ributzka — APPROVED** [review]
LGTM

**Pull request merged**